File: frontend/src/api/jobs.ts
import { api } from './client'
import type { JobListResponse, JobSubmitRequest, JobSubmitResponse, SlurmJob } from '@/types/api'

export interface JobOptions {
  partitions: string[]
  /**
   * GPU를 가진 노드가 있는 파티션. **`null`은 '모른다'**(노드 조회 실패)이고 빈 배열은
   * 'GPU가 없다'다 — 둘을 뭉뚱그리면 조회 실패가 제출을 막는다.
   */
  gpu_partitions: string[] | null
  accounts: string[]
  qos: string[]
  /**
   * 파티션별 **노드 한 대의 최대치**. 합계가 아니다 — `--cpus-per-task`·`--mem`은 노드
   * 경계를 넘지 못해서, 이보다 큰 값을 요청하면 Slurm이 제출을 거부한다(2014).
   * 비어 있으면 '모른다'이고, 그때는 화면이 상한을 강제하지 않는다.
   */
  capacity?: Record<string, { max_cpus_per_node: number; max_memory_mb: number }>
}

export const jobApi = {
  /** 제출 폼 선택지. 조회 실패한 항목은 빈 배열로 온다(slurmdbd만 끊긴 환경 등). */
  options: (cid: number) => api.get<JobOptions>(`/clusters/${cid}/job-options`),
  list: (cid: number, filters: Record<string, string | undefined> = {}) =>
    api.get<JobListResponse>(`/clusters/${cid}/jobs`, filters),
  history: (cid: number, filters: Record<string, string | undefined> = {}) =>
    api.get<JobListResponse>(`/clusters/${cid}/jobs/history`, filters),
  get: (cid: number, jobId: string) => api.get<SlurmJob>(`/clusters/${cid}/jobs/${jobId}`),
  submit: (cid: number, payload: JobSubmitRequest) =>
    api.post<JobSubmitResponse>(`/clusters/${cid}/jobs`, payload),
  previewScript: (cid: number, payload: JobSubmitRequest) =>
    api.post<{ script: string }>(`/clusters/${cid}/jobs/preview-script`, payload),
  cancel: (cid: number, jobId: string) => api.del(`/clusters/${cid}/jobs/${jobId}`),
  resubmit: (cid: number, jobId: string, overrides: Record<string, unknown> = {}) =>
    api.post<JobSubmitResponse>(`/clusters/${cid}/jobs/${jobId}/resubmit`, overrides),
  control: (cid: number, jobId: string, action: 'hold' | 'release' | 'priority', priority?: number) =>
    api.patch(`/clusters/${cid}/jobs/${jobId}`, { action, priority }),
}

File: backend/app/routers/jobs.py
"""Job 라우터 (api.md §5).

권한 표기가 `인증`인 엔드포인트도 **서버측에서 본인 범위로 강제**된다 —
관리자 여부에 따라 service가 스코프를 넓힌다.
"""

from typing import Annotated, Any

from fastapi import APIRouter, Depends

from app.core.deps import (
    AdminUser,
    ClientFactoryDep,
    CurrentUser,
    DbSession,
    SecretStoreDep,
    get_permission_cache,
    is_admin,
)
from app.core.redis_client import PermissionCache
from app.schemas.common import OkResponse
from app.schemas.job import (
    JobControlRequest,
    JobListResponse,
    JobResubmitRequest,
    JobSubmitRequest,
    JobSubmitResponse,
    ScriptPreview,
)
from app.services.cluster import ClusterService
from app.services.job import JobService, JobSpec

router = APIRouter(tags=["jobs"])


def _service(db: DbSession, secrets: SecretStoreDep, clients: ClientFactoryDep) -> JobService:
    return JobService(db, ClusterService(db, secrets=secrets, client_factory=clients))


JobServiceDep = Annotated[JobService, Depends(_service)]
PermCache = Annotated[PermissionCache, Depends(get_permission_cache)]


def _spec_from(payload: JobSubmitRequest) -> JobSpec:
    return JobSpec(**payload.model_dump())


@router.get("/clusters/{cid}/jobs", response_model=JobListResponse, summary="Job 목록")
def list_jobs(
    cid: int,
    user: CurrentUser,
    db: DbSession,
    cache: PermCache,
    service: JobServiceDep,
    state: str | None = None,
    partition: str | None = None,
    username: str | None = None,
) -> JobListResponse:
    cluster = service.clusters.get(cid)
    jobs = service.list_jobs(
        cluster,
        user=user,
        is_admin=is_admin(db, cache, user),
        state=state,
        partition=partition,
        username=username,
    )
    return JobListResponse(items=jobs, total=len(jobs))


@router.get("/clusters/{cid}/jobs/history", response_model=JobListResponse, summary="Job 이력")
def job_history(
    cid: int,
    user: CurrentUser,
    db: DbSession,
    cache: PermCache,
    service: JobServiceDep,
    start_time: str | None = None,
    end_time: str | None = None,
) -> JobListResponse:
    cluster = service.clusters.get(cid)
    jobs = service.history(
        cluster,
        user=user,
        is_admin=is_admin(db, cache, user),
        start_time=start_time,
        end_time=end_time,
    )
    return JobListResponse(items=jobs, total=len(jobs))


@router.post("/clusters/{cid}/jobs", response_model=JobSubmitResponse, summary="Job 제출")
def submit_job(
    cid: int, payload: JobSubmitRequest, user: CurrentUser, service: JobServiceDep
) -> JobSubmitResponse:
    cluster = service.clusters.get(cid)
    result = service.submit(cluster, _spec_from(payload), user=user)
    return JobSubmitResponse(**result)


@router.post(
    "/clusters/{cid}/jobs/preview-script",
    response_model=ScriptPreview,
    summary="제출 전 배치 스크립트 미리보기",
)
def preview_script(
    cid: int, payload: JobSubmitRequest, _: CurrentUser, service: JobServiceDep
) -> ScriptPreview:
    return ScriptPreview(script=service.build_script(_spec_from(payload)))


@router.get("/clusters/{cid}/jobs/{job_id}", response_model=dict, summary="Job 상세")
def get_job(
    cid: int,
    job_id: str,
    user: CurrentUser,
    db: DbSession,
    cache: PermCache,
    service: JobServiceDep,
) -> dict[str, Any]:
    cluster = service.clusters.get(cid)
    return service.get_job(cluster, job_id, user=user, is_admin=is_admin(db, cache, user))


@router.delete("/clusters/{cid}/jobs/{job_id}", response_model=OkResponse, summary="Job 취소")
def cancel_job(
    cid: int,
    job_id: str,
    user: CurrentUser,
    db: DbSession,
    cache: PermCache,
    service: JobServiceDep,
) -> OkResponse:
    cluster = service.clusters.get(cid)
    service.cancel(cluster, job_id, user=user, is_admin=is_admin(db, cache, user))
    return OkResponse(message="Job을 취소했습니다.")


@router.post(
    "/clusters/{cid}/jobs/{job_id}/resubmit",
    response_model=JobSubmitResponse,
    summary="Job 재제출",
)
def resubmit_job(
    cid: int,
    job_id: str,
    payload: JobResubmitRequest,
    user: CurrentUser,
    db: DbSession,
    cache: PermCache,
    service: JobServiceDep,
) -> JobSubmitResponse:
    cluster = service.clusters.get(cid)
    result = service.resubmit(
        cluster,
        job_id,
        user=user,
        is_admin=is_admin(db, cache, user),
        **payload.model_dump(exclude_none=True),
    )
    return JobSubmitResponse(**result)


@router.patch(
    "/clusters/{cid}/jobs/{job_id}", response_model=dict, summary="hold/release/우선순위 (ADMIN)"
)
def control_job(
    cid: int,
    job_id: str,
    payload: JobControlRequest,
    actor: AdminUser,
    service: JobServiceDep,
) -> dict[str, Any]:
    cluster = service.clusters.get(cid)
    return service.control(
        cluster, job_id, user=actor, action=payload.action, priority=payload.priority
    )
